### DISCUSSION TITLE
Auth and gate components

### DISCUSSION THREAD

**Discussion by binaryfire** [friendsofhyperf/components#277] [Ideas]
Hi @huangdijia

Great work on your components library! Would you consider making auth and gate components as well? I'm interested in migrating from Laravel to Hyperf for webapps, and the only thing that's missing is authentication (login / logout) and permissions (gates).

**Comment by huangdijia** [comment]
Hello, thank you for your affirmation and suggestions. I will carefully consider them. Here are some excellent components related to Auth. You can take a look at them first.

- [fx/hyperf-http-auth](https://github.com/nfangxu/hyperf-http-auth) 
- [96qbhy/hyperf-auth](https://github.com/qbhy/hyperf-auth) 
- [hyperf-ext/jwt](https://github.com/hyperf-ext/jwt) 
- [hyperf-ext/auth](https://github.com/hyperf-ext/auth) 
- [donjan-deng/hyperf-casbin](https://github.com/donjan-deng/hyperf-casbin) 
- [Casbin](https://casbin.org/docs/zh-CN/overview)

**Reply by binaryfire** [reply]
Thanks for your suggestions @huangdijia. I checked the components you recommended. Casbin is too complex for what I need, so I'll write my own helper for permissions for now.

All the authentication components are abandoned, except [96qbhy]([96qbhy](https://github.com/qbhy/hyperf-auth)). I will try that one and see if it works.

**Reply by hsluoyz** [reply]
@huangdijia you can also consider Casdoor: https://casdoor.org/

**Comment by huangdijia** [comment]
I took the time to adapt auth to Hyperf, and you are welcome to participate.

**Reply by binaryfire** [reply]
@huangdijia Awesome! I'd definitely like to participate.

**Reply by binaryfire** [reply]
@huangdijia Secure website login also needs CSRF protection. None of the auth components in that list have a CSRF feature, so they are all insecure!

Here is some info on CSRF: 
https://laravel.com/docs/10.x/csrf
https://github.com/slimphp/Slim-Csrf
https://github.com/muhamed-didovic/csrf
https://symfony.com/doc/current/security/csrf.html

(CSRF protection is required to securely submit any session-authenticated POST/PUT/PATCH requests on the website. Login form, contact form, API endpoints etc...)

**Reply by huangdijia** [reply]
Can you create a pr, submit this feature to the https://github.com/hyperf/hyperf？ We'll finish it together.

**Reply by binaryfire** [reply]
@huangdijia I'd love to help but unfortunately my coding skills aren't good enough to create components from scratch. I can only offer suggestions and help with testing, feedback and minor PRs.

**Comment by binaryfire** [comment]
Closing. `hyperf/session` already includes most of the building blocks needed for implementing auth. Including [CSRF token generation](https://github.com/hyperf/hyperf/blob/02f57186eb8693b32021c41c4d3ac70ff55e4303/src/session/src/Session.php#L265)
